airbyte-integrations/connectors/source-mysql-v2/src/main/kotlin/io/airbyte/integrations/source/mysql/MysqlJdbcStreamStateValue.kt

```diff
@@ -0,0 +1,92 @@
+    @JsonProperty("cursor_record_count") val cursorRecordCount: Int = 0,
+    @JsonProperty("pk_name") val pkName: String? = null,
+    @JsonProperty("pk_value") val pkValue: String? = null,
+    @JsonProperty("incremental_state") val incrementalState: JsonNode? = null,
+) {
+    companion object {
+        /** Value representing the completion of a FULL_REFRESH snapshot. */
+        val snapshotCompleted: OpaqueStateValue
+            get() = Jsons.valueToTree(MysqlJdbcStreamStateValue(stateType = "primary_key"))
+
+        /** Value representing the progress of an ongoing incremental cursor read. */
+        fun cursorIncrementalCheckpoint(
+            cursor: Field,
+            cursorCheckpoint: JsonNode,
+            stream: Stream,
+        ): OpaqueStateValue {
+            return Jsons.valueToTree(
+                MysqlJdbcStreamStateValue(
+                    cursorField = listOf(cursor.id),
+                    cursors = cursorCheckpoint.asText(),
+                    streamName = stream.name,
+                    streamNamespace = stream.namespace!!
+                )
+            )
+        }
+
+        /** Value representing the progress of a ongoing snapshot not involving cursor columns. */
+        fun snapshotCheckpoint(
+            primaryKey: List<Field>,
+            primaryKeyCheckpoint: List<JsonNode>,
+        ): OpaqueStateValue {
+            val primaryKeyField = primaryKey.first()
+            return Jsons.valueToTree(
+                MysqlJdbcStreamStateValue(
+                    pkName = primaryKeyField.id,
+                    pkValue = primaryKeyCheckpoint.first().asText(),
+                    stateType = StateType.PRIMARY_KEY.stateType,
+                )
+            )
+        }
+
+        /** Value representing the progress of an ongoing snapshot involving cursor columns. */
+        fun snapshotWithCursorCheckpoint(
+            primaryKey: List<Field>,
+            primaryKeyCheckpoint: List<JsonNode>,
+            cursor: Field,
+            stream: Stream
+        ): OpaqueStateValue {
+            val primaryKeyField = primaryKey.first()
+            return Jsons.valueToTree(
+                MysqlJdbcStreamStateValue(
+                    pkName = primaryKeyField.id,
+                    pkValue = primaryKeyCheckpoint.first().asText(),
+                    stateType = StateType.PRIMARY_KEY.stateType,
+                    incrementalState =
+                        Jsons.valueToTree(
+                            MysqlJdbcStreamStateValue(
+                                cursorField = listOf(cursor.id),
+                                streamName = stream.name,
+                                streamNamespace = stream.namespace!!
+                            )
+                        ),
+                )
+            )
+        }
+    }
+}
+
+enum class StateType(val stateType: String) {
+    PRIMARY_KEY("primary_key"),
+    CURSOR_BASED("cursor_based"),
+}
```

Comment:
non blocking comment: FYI this could have been
```
enum class StateType {
    PRIMARY_KEY,
    CURSOR_BASED;

    val stateType: String
        get() = name.lowercase()
}

Reply:
don't bother making this change, this is just an informational comment
your code is fine